mne/time_frequency/psd.py

```diff
@@ -156,44 +162,68 @@ def compute_epochs_psd(epochs, picks=None, fmin=0, fmax=np.inf, tmin=None,
-        time_mask = _time_mask(epochs.times, tmin, tmax)
-    else:
-        time_mask = slice(None)
-
-    data = epochs.get_data()[:, picks][..., time_mask]
-    if proj:
-        proj, _ = make_projector_info(epochs.info)
-        if picks is not None:
-            data = np.dot(proj[picks][:, picks], data)
+    from .multitaper import multitaper_psd
+    from ..epochs import _BaseEpochs
+    if isinstance(epochs, _BaseEpochs):
+        Fs = epochs.info['sfreq']
+        if picks is None:
+            picks = pick_types(epochs.info, meg=True, eeg=True, ref_meg=False,
+                               exclude='bads')
+
+        if tmin is not None or tmax is not None:
+            time_mask = _time_mask(epochs.times, tmin, tmax)
         else:
-            data = np.dot(proj, data)
-
-    logger.info("Effective window size : %0.3f (s)" % (n_fft / float(Fs)))
-
-    freqs = np.arange(n_fft // 2 + 1, dtype=float) * (Fs / n_fft)
-    freq_mask = (freqs >= fmin) & (freqs <= fmax)
-    freqs = freqs[freq_mask]
-    psds = np.empty(data.shape[:-1] + (freqs.size,))
-
-    parallel, my_pwelch, n_jobs = parallel_func(_pwelch, n_jobs=n_jobs,
-                                                verbose=verbose)
-
-    for idx, fepochs in zip(np.array_split(np.arange(len(data)), n_jobs),
-                            parallel(my_pwelch(epoch, noverlap=n_overlap,
-                                               nfft=n_fft, fs=Fs,
-                                               freq_mask=freq_mask,
-                                               welch_fun=welch)
-                                     for epoch in np.array_split(data,
-                                                                 n_jobs))):
-        for i_epoch, f_epoch in zip(idx, fepochs):
-            psds[i_epoch, :, :] = f_epoch
+            time_mask = slice(None)
+
+        data = epochs.get_data()[:, picks][..., time_mask]
+        if proj:
+            proj, _ = make_projector_info(epochs.info)
+            if picks is not None:
+                data = np.dot(proj[picks][:, picks], data)
+            else:
+                data = np.dot(proj, data)
+    elif Fs is None:
+        raise ValueError('Must give sampling frequency (Fs) if array provided')
+    else:
+        data = np.atleast_3d(epochs)
+
+    n_epochs, n_channels, n_times = data.shape
+    if method == 'welch':
+        n_fft, n_overlap = _check_nfft(n_times, n_fft, n_overlap)
+        n_fft = int(n_fft)
+        logger.info("Effective window size : %0.3f (s)" % (n_fft / float(Fs)))
+
+        freqs = np.arange(n_fft // 2 + 1, dtype=float) * (Fs / n_fft)
+        freq_mask = (freqs >= fmin) & (freqs <= fmax)
+        freqs = freqs[freq_mask]
+        psds = np.empty(data.shape[:-1] + (freqs.size,))
+
+        parallel, my_pwelch, n_jobs = parallel_func(_pwelch, n_jobs=n_jobs,
+                                                    verbose=verbose)
+
+        for idx, fepochs in zip(np.array_split(np.arange(len(data)), n_jobs),
```

Comment:
yeah, it was cryptic to me too haha. A lot of this code is stuff that was already there. I'll try to implement what you've described.